<h2> Introduction to using Jupyter </h2>

This notebook is meant to give you an introduction to using Jupyter. This is a programming platform that we'll be using throughout Math 032 this semester; although you won't be doing too much coding, there will be assignments where you need to write or modify code snippets so that we can use Jupyter's computational abilities to answer questions in probability and statistics. Since this is a heavily applied course with a lot of real-world applications, it's especially important to be able to use computational tools to solve problems that are beyond the scope of pen-and-paper. 

Our Jupyter notebooks mix markdown (that's what this cell is -- really, it's just fancy text with a little bit of formatting!) and Python code. You can execute the code in any cell by either clicking the `run` button above this block, or by pressing `Ctrl-Enter` or `Cmd-Enter`. In a typical notebook, you'll be given some templates or code to modify to a particular usage; you shouldn't have to write things from scratch, but rather edit previously given material.

This notebook has a couple of questions that introduce you to the process of using Python code to answer questions, along with the particular syntax. If you're interested in learning more about the details of coding in Python, [this tutorial](https://www.w3schools.com/python/) is a great starting point.

Let's get started!

<h3> Part 1: Computing a sum with Python </h3>

In previous courses, you probability computed sums with sigma notation; for example,
$$\sum_{n = 1}^{1000} \frac{1}{n} = \frac 1 1 + \frac 1 2 + \frac 1 3 + \cdots \frac{1}{1000}.$$
These can get extremely cumbersome to compute by hand, but are easily handled by a computer using a [for loop](https://www.w3schools.com/python/python_for_loops.asp). The Python code that will compute this sum is below, with a lot of comments. Comments are placed on lines using \#, and are not interpreted as code.

In [1]:
# This is a comment!

# First, we set a variable to be 0. This is 
# our initial sum.
s = 0

# Now we do a for loop. We take a variable n that
# runs from 1 to 1000. Python does this using the 
# keyword range: range(a, b) is a list of numbers
# from a to b - 1, inclusive. Note it doesn't 
# include b itself, so we have to modify this.

for n in range(1, 1000 + 1):
    # In each loop, we add the next piece, which 
    # is just 1/n. We keep all the code in the
    # loop indented so that Python knows we're 
    # still in the loop.
    s = s + 1/n

# Tell us what the sum is:
print(s)




7.485470860550343


So the sum of the first $1000$ numbers is about $7.49$. Now it's your turn; add any new code that you need to **answer the following two questions**.

* Compute $\sum_{n = 1}^{100} n^3$. Note that in Python, $a^b$ is written as `a**b`.

* Estimate $\sum_{n = 1}^{100} \frac{1}{n^2}$, $\sum_{n = 1}^{1000} \frac{1}{n^2}$, and $\sum_{n = 1}^{10^6} \frac{1}{n^2}.$

Once you've got results, put them in the next block.

<h3><i> Put your answers for Part 1 here! </i> </h3>

<h3> Part 2: Generating random numbers with Python </h3>

One of the main uses we'll have for Python this semester is generating random data; we can use Python to simulate rolling dice, drawing cards, building random data from a distribution, and many more things. A lot of this will use Python's [random](https://www.geeksforgeeks.org/python-random-function/) module, so this example will involve importing and running code from this module.

Here's how we can use Python to flip a coin: generate a "uniformly distributed" random number between $0$ and $1$ (we'll say what this means in Chapter 5). If it's bigger than $0.5$, call it Heads and if it's smaller than $0.5$, call it Tails. Let's count how many heads we get if we flip $100$ coins:

In [3]:
# First we need to tell Python where to find 
# the random number module:

from random import random

# Now we flip the coin 100 times and record
# the number of heads. Start by setting it
# to zero.
heads = 0

# Flip! Saying range(100) is shorthand for
# range(0, 100) = [0, 1, 2, ..., 99].
for n in range(100):
    # Gen
    flip = random()

    # Now check if it's bigger than 0.5:
    if flip > 0.5:
        # Note that we're inside an if
        # block now. We indent even more.
        heads = heads + 1
        # We could also write heads += 1 
        # for this. It just records a heads.

# Now report how many heads we got. The f'...'
# prints the ... with a bit of formatting.
print(f'Number of heads = {heads}')

Number of heads = 56


The first time I ran this, I got 47 heads; the second gave me 51, and the third gave me 56.

Now it's your turn! Add new blocks to **answer the following two questions**:

* Suppose you flip $1000$ coins and record the number of heads. Carry this experiment out $10$ times; what's the largest number of heads that you recorded in those ten runs? (You can do this by hand or automate it with another for loop).

* How likely do you think it is to see a result of at least $600$ heads out of $1000$ flips? Do you think it's somewhat unlikely, very unlikely, or extremely unlikely?

<h3><i> Put your answers for Part 2 here!</i></h3>

<h3> Part 3: Submitting this to Gradescope </h3>

Once you've finished modifying your notebook and answering the questions, you'll need to submit it to Gradescope along with your other homework. To do this, generate a pdf file by clicking `File -> Save and Export Notebook as... -> PDF`. Then upload that PDF to Gradescope and submit it to the assignment `Jupyter 1 - Introduction`. As always -- if you have any questions or run into any issues you can
* ask during discussion,
* email your TA or instructor,
* or bring them to student hours!